libs/stokesian_dynamics/sd.hpp

```diff
@@ -0,0 +1,1478 @@
+#include <cstdio>
+#include <vector>
+
+#include "thrust_wrapper.hpp"
+#include "device_matrix.hpp"
+#include "multi_array.hpp"
+
+#ifdef __CUDACC__
+#include <curand_kernel.h>
+#else
+#include <Random123/philox.h>
+
+typedef r123::Philox2x64 RNG;
+#endif
+
+
+#ifdef __CUDACC__
+#define DEVICE_FUNC __host__ __device__
+#define KERNEL_ABORT asm("trap;")
+#else
+#define DEVICE_FUNC
+#define KERNEL_ABORT abort()
+#endif
+
+/** \file
+ *  This file contains computations required for Stokesioan Dynamics, namely
+ *  the particle's translational and angular velocities are computed from the 
+ *  particle's positions and radii and the forces and torques that are acting
+ *  on the particles.
+ *
+ *  In its current implementation, the neccessary subroutines are implemented
+ *  as functors. The main functor is called solver, which wraps up the
+ *  functionality of all the other functors in this file. For details, please
+ *  see the description of the solver functor.
+ *
+ *  For details on the thermalization, please see the description of the
+ *  thermalizer functor.
+ *
+ *  The general Stokesian Dynamics method also allows to incorporate one
+ *  stress tensor per particle which gives the stress that is imposed on a
+ *  particle by the surrounding fluid (such as to deform the particle).
+ *  However, ESPResSo makes no use of that, leaving only forces and torques
+ *  to be passed to the SD routine.
+ *
+ *  To understand the Stokesian Dynamics method, it is recommended to read
+ *  Durlofski et al. (1987), because it describes the method that is
+ *  implemented here, apart from the thermalization.
+ *  (F-T-version or F-T-S-version?)
+ *      "Dynamic simulation of hydrodynamically interacting suspensions"
+ *      https://core.ac.uk/download/pdf/4895234.pdf
+ *
+ *  All references to formulae in this file refer to the aforementioned paper,
+ *  unless otherwise noted.
+ *
+ *  As they state in the paragraph below equation (2.17) it may be useful to
+ *  also read Jeffrey et al. (1984) to better understand the notation used to
+ *  compute the grand mobility matrix in this Stokesian Dynamics method.
+ *      "Calculation of the resistance and mobility functions for two unequal
+ *      rigid spheres in low-Reynolds-number flow"
+ *      https://doi.org/10.1017/S0022112084000355
+ *
+ *
+ *  Also, the figures in
+ *      https://doi.org/10.1016/j.jcp.2015.01.019
+ *  might help with an intuitive understanding of dipole, stokeslet and rotlet
+ *  flow (missing out only on the stresslet flow).
+ *
+ */
+
+
+namespace sd {
+
+enum flags {
```

Comment:
Please use C++11 enum classes where possible.

Reply:
I read about C++11 enum classes and I am not sure if I can replace this. The names that are defined within this enumeration are used for checking individual bits in a flags register. Maybe "enumeration" is not the right term. What about
```suggestion
namespace flags {
    NONE = 0,
    SELF_MOBILITY = 1 << 0,
    PAIR_MOBILITY = 1 << 1,
    LUBRICATION = 1 << 2,
    FTS = 1 << 3,
```

Reply:
`namespace flags` will lead to a compiler error, as the flags will be considered undeclared variables. The idea behind rewriting `enum flags` as `enum class flags` is that you get namespaced enum values, e.g. `SELF_MOBILITY` is only accessible through `flags::SELF_MOBILITY`. Such enum values are passed to functions as type `flags` instead of `int`, which helps generating compiler errors when you accidentally add an enum value to some other `int`. The drawback as you've noted, is that we cannot do bitwise operations anymore without a lot of boilerplate code.  What you can do is write `namespace flags { enum flags {...};} ` which will scope them while still allowing bitwise operations.

Reply:
Thanks for the clarification and suggestion.

Reply:
Why don't you just leave the `enum` as it is?  It is scoped already in the `sd` namespace, so an additional indirection is really not needed.

Reply:
That's actually true. Well, It doesn't harm either, I guess.